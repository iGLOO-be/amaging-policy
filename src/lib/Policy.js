
import isObject from 'lodash/isObject'
import filter from 'lodash/filter'
import extend from 'lodash/extend'
import toArray from 'lodash/toArray'
import Store from 'dottystore'
import validators from './validators'

/*
  Policy sample:

  { "expiration": "2007-12-01T12:00:00.000Z",
    "conditions": [
      ["starts-with", "$key", "user/eric/"],
      ["range-length", "$key", 10, 50],
      {"success_action_redirect": "http://johnsmith.s3.amazonaws.com/new_post.html"},
    ]
  }

*/

class PolicyError extends Error {
  constructor (type, data) {
    super(...arguments)

    this.type = type
    this.data = data
    this.name = 'PolicyError'
    this.message = (() => {
      switch (this.type) {
        case 'INVALID_KEY':
          return `Invalid value for key: ${this.data.key}`
        case 'INVALID_VALIDATOR_NAME':
          return `Invalid policy validator name: ${this.data.validator}`
      }
    })()
  }
}

class Condition {
  constructor (key, validatorName, validator, validatorArgs) {
    this.key = key
    this.validatorName = validatorName
    this.validator = validator
    this.validatorArgs = validatorArgs
  }
  valid (value) { return this.validator.apply(null, [value].concat(toArray(this.validatorArgs))) }
}

class Policy extends Store {
  static initClass () {
    this.validators = {}
  }
  static registerValidators (validators) {
    return (() => {
      const result = []
      for (let validatorName in validators) {
        result.push(Policy.registerValidator(validatorName, validators[validatorName]))
      }
      return result
    })()
  }
  static registerValidator (validatorName, validator) {
    Policy.validators[validatorName] = validator
  }
  static getValidator (validatorName) {
    return Policy.validators[validatorName]
  }

  constructor (policy) {
    super(...arguments)

    this._parsePolicy(policy)
  }

  set (key, value) {
    const conditions = this.getConditionForKey(key)

    for (let cond of Array.from(conditions || [])) {
      if (!cond.valid(value)) {
        throw new PolicyError('INVALID_KEY', {key})
      }
    }

    return super.set(...arguments)
  }

  _parsePolicy (policy = {}) {
    let key
    const data = {}
    const conditions = []

    for (let pol of Array.from(policy.conditions || [])) {
      if (Array.isArray(pol)) {
        const validatorName = pol[0]
        key = pol[1].replace(/^$/, '')
        const validatorArgs = pol.slice(2)

        const validator = Policy.getValidator(validatorName)
        if (!validator) {
          throw new PolicyError('INVALID_VALIDATOR_NAME', {validator})
        }

        const condition = new Condition(key, validatorName, validator, validatorArgs)
        conditions.push(condition)
      } else if (isObject(pol)) {
        extend(data, pol)
      }
    }

    this.conditions = conditions

    return (() => {
      const result = []
      for (key in data) {
        result.push(this.set(key, data[key]))
      }
      return result
    })()
  }

  getConditionForKey (key) {
    return filter(this.conditions, {key})
  }
}
Policy.initClass()

Policy.registerValidators(validators)

export default Policy
