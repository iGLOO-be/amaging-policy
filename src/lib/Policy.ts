
import { Store } from "dottystore";
import extend from "lodash/extend";
import isObject from "lodash/isObject";
import { Condition } from "./Condition";
import { PolicyError } from "./PolicyError";
import { ValidatorName, validators } from "./validators";

/*
  Policy sample:s

  { "expiration": "2007-12-01T12:00:00.000Z",
    "conditions": [
      ["starts-with", "$key", "user/eric/"],
      ["range-length", "$key", 10, 50],
      {"success_action_redirect": "http://johnsmith.s3.amazonaws.com/new_post.html"},
    ]
  }

*/

export interface IPolicyObject {
  conditions?: IPolicyCondition[];
}
export type IPolicyCondition = [ValidatorName, string, ...any[]] | IPolicyData;
interface IPolicyData {
  [key: string]: string;
}

export
class Policy extends Store {
  public conditions: Condition[];

  constructor(policy?: IPolicyObject) {
    super();

    this.conditions = [];
    this._parsePolicy(policy);
  }

  public set(key: string, value: any) {
    const conditions = this.getConditionsForKey(key);

    for (const cond of Array.from(conditions || [])) {
      if (!cond.valid(value)) {
        throw new PolicyError("INVALID_KEY", {key});
      }
    }

    return super.set(key, value);
  }

  public _parsePolicy(policy: IPolicyObject = {}) {
    let key;
    const data: IPolicyData = {};
    const conditions = [];

    for (const pol of Array.from(policy.conditions || [])) {
      if (Array.isArray(pol)) {
        const validatorName = pol[0];
        key = pol[1].replace(/^$/, "");
        const validatorArgs = pol.slice(2);

        const validator = validators[validatorName];
        if (!validator) {
          throw new PolicyError("INVALID_VALIDATOR_NAME", {validator});
        }

        const condition = new Condition(key, validatorName, validator, validatorArgs);
        conditions.push(condition);
      } else if (isObject(pol)) {
        extend(data, pol);
      }
    }

    this.conditions = conditions;

    Object.keys(data).forEach((dataKey) => {
      this.set(dataKey, data[dataKey]);
    });
  }

  public getConditionsForKey(key: string): Condition[] {
    return this.conditions.filter((cond) => cond.key === key);
  }
}
