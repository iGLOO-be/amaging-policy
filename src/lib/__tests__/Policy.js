
import Policy from '../Policy'

describe('Policy', () => {
  it('Can be instanciated with all arguments', () => {
    const policy = new Policy()
    expect(policy).toBeTruthy()
  })

  it('Throws an error if policy data do not match policy conditions', function () {
    expect(function () {
      const policy = new Policy({
        conditions: [
          {'foo': 'bar'},
          ['eq', 'foo', 'baz']
        ]
      })
      console.log(policy)
    }).toThrow('Invalid value for key: foo')
  })

  it('Not throws an error if policy data do not match policy conditions', function () {
    const policy = new Policy({
      conditions: [
        {'foo': 'bar'},
        ['eq', 'foo', 'bar']
      ]
    })
    expect(policy.get('foo')).toEqual('bar')
  })

  it('Throws an error if policy data do not match the second policy conditions', function () {
    expect(function () {
      const policy = new Policy({
        conditions: [
          {'foo': 'bar'},
          ['eq', 'foo', 'bar'],
          ['eq', 'foo', 'baz']
        ]
      })
      console.log(policy)
    }).toThrow('Invalid value for key: foo')
  })

  describe('Policy::set', function () {
    it('Can set a value without validations', function () {
      const policy = new Policy()
      policy.set('foo', 'bar')
      expect(policy.get('foo')).toEqual('bar')
    })

    it('Can set a value with validations', function () {
      const policy = new Policy({
        conditions: [
          ['eq', 'foo', 'bar']
        ]
      })
      expect(policy.get('foo')).toEqual(undefined)
      policy.set('foo', 'bar')
      expect(policy.get('foo')).toEqual('bar')
    })

    it('Throws an error if set a value breaking a validation', function () {
      const policy = new Policy({
        conditions: [
          ['eq', 'foo', 'bar']
        ]
      })
      expect(policy.get('foo')).toEqual(undefined)
      expect(() => policy.set('foo', 'baz')).toThrow('Invalid value for key: foo')
      expect(policy.get('foo')).toEqual(undefined)
    })

    it('Throws an error if set a value breaking the second validation', function () {
      const policy = new Policy({
        conditions: [
          ['eq', 'foo', 'bar'],
          ['eq', 'foo', 'baz']
        ]
      })
      expect(policy.get('foo')).toEqual(undefined)
      expect(() => policy.set('foo', 'bar')).toThrow('Invalid value for key: foo')
      expect(policy.get('foo')).toEqual(undefined)
    })
  })

  describe('Policy::getConditionForKey', function () {
    it('Can get conditions for a key', function () {
      const policy = new Policy({
        conditions: [
          ['eq', 'foo', 'bar'],
          ['eq', 'foo', 'baz'],
          ['eq', 'bar', 'baz']
        ]
      })
      expect(policy.getConditionForKey('foo')).toMatchSnapshot()
    })
  })
})
