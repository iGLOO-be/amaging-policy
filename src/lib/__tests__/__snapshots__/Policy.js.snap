// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policy Policy::getConditionsForKey Can get conditions for a key 1`] = `
Array [
  Condition {
    "key": "foo",
    "validator": [Function],
    "validatorArgs": Array [
      "bar",
    ],
    "validatorName": "eq",
  },
  Condition {
    "key": "foo",
    "validator": [Function],
    "validatorArgs": Array [
      "baz",
    ],
    "validatorName": "eq",
  },
]
`;
